analyzers/src/SonarAnalyzer.Common/Json/Parsing/LexicalAnalyzer.cs

```diff
@@ -0,0 +1,313 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SonarAnalyzer.Json.Parsing
+{
+    internal class LexicalAnalyzer
+    {
+        private readonly List<string> lines = new List<string>();
+        private int line;
+        private int position = -1;
+
+        public object Value { get; private set; }
+        public LinePosition LastStart { get; private set; }
+        public LinePosition LastEnd { get; private set; }
+        private char CurrentChar => lines[line][position];
+        private bool CanReadCurrentChar => line < lines.Count - 1 || (line == lines.Count - 1 && position < lines[line].Length);
+        private string AtLocationSuffix => $" at line {LastStart.Line + 1} position {LastStart.Character + 1}";
+
+        public LexicalAnalyzer(string source)
+        {
+            var sb = new StringBuilder();
+            foreach (var c in source.Replace("\r\n", "\n"))
+            {
+                if (c == '\n'
+                    || c == '\r'
+                    || (char.GetUnicodeCategory(c) is var category && (category == UnicodeCategory.LineSeparator || category == UnicodeCategory.ParagraphSeparator)))
+                {
+                    lines.Add(sb.ToString());
+                    sb.Clear();
+                }
+                else
+                {
+                    sb.Append(c);
+                }
+            }
+            lines.Add(sb.ToString());
```

Comment:
Using `Span<string>` instead should save some memory, but I'm not sure if we can use it in our codebase

Reply:
What are the objections? (Just curious).

Reply:
There are no objections, as it is a clear winner. But we target `net46` as a technical limitation.

Reply:
Oops, fair point.